force/main/default/classes/StubProvider.cls

```diff
@@ -23,46 +23,17 @@ public class StubProvider implements System.StubProvider
 	}
 
 	public void stub(Object stubbedObject, String stubbedMethodName, List<Type> listOfParamTypes, List<Object> listOfArgs, Object returnValue)
-	{
-		returnValuesByMethod.put(new Method(stubbedObject, stubbedMethodName, listOfParamTypes, listOfArgs), returnValue);
+	{	
+		returnValuesByMethod.put(generateHash(stubbedObject, stubbedMethodName, listOfParamTypes, listOfArgs), returnValue);
 	}
 
-	private class Method
-	{
-		private final Object stubbedObject;
-		private final String stubbedMethodName;
-		private final List<Type> listOfParamTypes;
-		private final List<Object> listOfArgs;
-
-		public Method(Object stubbedObject, String stubbedMethodName, List<Type> listOfParamTypes, List<Object> listOfArgs)
-		{
-			this.stubbedObject = stubbedObject;
-			this.stubbedMethodName = stubbedMethodName;
-			this.listOfParamTypes = listOfParamTypes;
-			this.listOfArgs = listOfArgs;
-		}
-
-		public Boolean equals(Object other)
-		{
-			if (other instanceof Method)
-			{
-				Method that = (Method)other;
-				return this.stubbedObject === stubbedObject
-					&& this.stubbedMethodName == stubbedMethodName
-					&& this.listOfParamTypes == listOfParamTypes
-					&& this.listOfArgs == listOfArgs;
-			}
-
-			return false;
-		}
-
-		public Integer hashCode()
-		{
-			String description = ''
-				+ this.stubbedMethodName
-				+ this.listOfParamTypes;
-
-			return description.hashCode();
-		}
+	private Integer generateHash(Object stubbedObject, String stubbedMethodName, List<Type> listOfParamTypes, List<Object> listOfArgs) {
+		String objectType = ('' + stubbedObject).split(':')[0];
```

Comment:
Stubbing now applies to all instances of a given SObject type. Previously it was instance specific, because the equals checked for referential equality.
Also, now we're reliant on the Object's toString method, which can be overridden, or the result of which can change during the course of a test.

I don't think this is the right fix :/